### PR TITLE
Refactor response

### PR DESCRIPTION
**Pull request by mgautierfr**
With the new libzim API, it would be not possible to have an "not constructed" `zim::Entry` or `zim::Item`.
However, the server's `Response` has a `kiwix::Entry` (which is turn, wrap a `zim::Entry`).
So, with the actual implementation, it would be impossible to instanciate a `Response` without having a `zim::Entry` (with the new implementation).

This PR splits the different kinds of response in different classes.
So only the `EntryResponse` will need to have a entry.

The code of this PR is iso-functional with previous code.

Questions (@veloman-yunkan you may help here) :
I'm not sure of how to handle the `ResponseMode`.
`ResponseMode::ERROR_RESPONSE` is used only if we cannot parse the byteRange when we try to create an EntryResponse. In this case, we don't set the headers in the `send` method.
I don't know why we must not set those headers in this case. 
If it is because this in an error, should we avoid to set the header for 500 (or 404) also ?

The same way, for 304 response, I use a ContentResponse to let the `ContentResponse::create_mhd_response` set the headers (especially MHD_HTTP_HEADER_VARY).
If I use a `Response`, `MHD_HTTP_HEADER_VARY` is missing and test IfNoneMatchRequestsWithMatchingETagResultIn304Responses fails.

### CODE REVIEW

**Comment by veloman-yunkan** [comment]
> I'm not sure of how to handle the `ResponseMode`.
> `ResponseMode::ERROR_RESPONSE` is used only if we cannot parse the byteRange when we try to create an EntryResponse. In this case, we don't set the headers in the `send` method.
> I don't know why we must not set those headers in this case.
> If it is because this in an error, should we avoid to set the header for 500 (or 404) also ?

I wrote the code for generating a 416 error response to an invalid byte range request in an ad-hoc manner. I don't think that I intended to omit any headers, I just included the only header required by the spec for that response.

**Comment by mgautierfr** [comment]
I have remove the `responseMode` and create a specific factory for 304. My two questions are answered.

File: src/server/response.cpp [hunk 1]
```diff
@@ -275,14 +274,12 @@ MHD_Result Response::send(const RequestContext& request, MHD_Connection* connect
...
-    const std::string etag = m_etag.get_etag();
-    if ( ! etag.empty() )
-        MHD_add_response_header(response, MHD_HTTP_HEADER_ETAG, etag.c_str());
-  }
+  MHD_add_response_header(response, "Access-Control-Allow-Origin", "*");
+  MHD_add_response_header(response, MHD_HTTP_HEADER_CACHE_CONTROL,
+    m_etag.get_option(ETag::CACHEABLE_ENTITY) ? "max-age=2723040, public" : "no-cache, no-store, must-revalidate");
+  const std::string etag = m_etag.get_etag();
+  if ( ! etag.empty() )
+    MHD_add_response_header(response, MHD_HTTP_HEADER_ETAG, etag.c_str());
```
Comment (veloman-yunkan) [inline src/server/response.cpp]: Now I see what you meant in your question about excluded headers. In the morning I had looked in the wrong place (`Response::create_error_response()` and its call site). Should have read your question twice. I definitely made this change intentionally, but I am not sure if it is really required or not. At least my unit test for 416 responses doesn't validate the absence of those headers.

File: src/server/internalServer.cpp [hunk 2]
```diff
@@ -791,11 +782,11 @@ std::unique_ptr<Response> InternalServer::handle_content(const RequestContext& r
...
   if (m_verbose.load()) {
     printf("Found %s\n", entry.getPath().c_str());
-    printf("mimeType: %s\n", response->get_mimeType().c_str());
+    printf("mimeType: %s\n", entry.getMimetype().c_str());
   }
 
-  if (response->get_mimeType().find("text/html") != string::npos)
-    response->set_taskbar(bookName, reader->getTitle());
+  if (entry.getMimetype().find("text/html") != string::npos)
+    dynamic_cast<ContentResponse*>(response.get())->set_taskbar(bookName, reader->getTitle());
```
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/server/internalServer.cpp]: I now `dynamic_cast` with reference and catch the exception. There is no more test on the mimeType.

**Comment by mgautierfr** [comment]
The last push is a rebase on master, fix of the dynamic_cast and remove of `m_compress` member.

It remains the 416's headers point but I don't know if it is a blocker or not. Tests are not failing at least.

If it is ok for you @veloman-yunkan, it would be nice to merge this to unlock other PR on libzim_next and avoid potential conflict with future PR.

**Comment by veloman-yunkan** [comment]
> It remains the 416's headers point but I don't know if it is a blocker or not. Tests are not failing at least.

The headers of the 416 error response are not a blocker for me.

**Pull request merged**